<a href="https://colab.research.google.com/github/ChungYuanHsu/Python-and-ML_homework/blob/master/PythonML_HWQ9_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color=1978DD> HWQ9-1 使用CNN來做圖片辨識 </font> 
## <font color=FC8600> 重建課堂中model </font> 
  ###  <font color=2214AF >  $\;\;$  載入需要用到的套件  </font>

In [1]:
%env KERAS_BACKEND=tensorflow
%matplotlib inline

import numpy as np
import matplotlib.pyplot as plt

env: KERAS_BACKEND=tensorflow


###  <font color=2214AF >  $\;\;$ 載入手寫數字的資料  </font>

In [2]:
from keras.datasets import mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

Using TensorFlow backend.


###  <font color=2214AF >  $\;\;$ 將每張圖片原本的資料 (28,28) 轉成 (28,28,1)</font>

In [0]:
x_train = x_train.reshape(60000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)

###  <font color=2214AF >  $\;\;$ 輸出分類數目為0-9共10種  </font>

In [0]:
from keras.utils import np_utils
y_train = np_utils.to_categorical(y_train, 10)
y_test = np_utils.to_categorical(y_test, 10)

###  <font color=2214AF >  $\;\;$ 把資料調整到0~1的範圍之內，使CNN計算時較穩定 </font>

In [0]:
x_train = x_train/255
x_test = x_test/255

###  <font color=2214AF >  $\;\;$ 載入套件</font>

In [0]:
from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten
from keras.layers import Conv2D, MaxPool2D
from keras.optimizers import SGD

###  <font color=2214AF >  $\;\;$ 建立神經網路</font>

In [7]:
model = Sequential()

model.add(Conv2D(4, (5, 5), padding='same', input_shape=(28, 28, 1)))
model.add(Activation('relu'))

model.add(MaxPool2D(pool_size=(2,2)))

model.add(Conv2D(8, (5, 5), padding='same'))
model.add(Activation('relu'))

model.add(MaxPool2D(pool_size=(2,2)))

model.add(Conv2D(16, (5, 5), padding='same'))
model.add(Activation('relu'))

model.add(MaxPool2D(pool_size=(2,2)))

model.add(Flatten())
model.add(Dense(9))
model.add(Activation('relu'))

model.add(Dense(10))
model.add(Activation('softmax'))

Instructions for updating:
Colocations handled automatically by placer.


###  <font color=2214AF >  $\;\;$ compile</font>

In [0]:
model.compile(loss='mse', optimizer=SGD(lr=0.07), metrics=['accuracy'])

###  <font color=2214AF >  $\;\;$ summary</font>

In [9]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 28, 28, 4)         104       
_________________________________________________________________
activation_1 (Activation)    (None, 28, 28, 4)         0         
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 14, 14, 4)         0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 14, 14, 8)         808       
_________________________________________________________________
activation_2 (Activation)    (None, 14, 14, 8)         0         
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 7, 7, 8)           0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 7, 7, 16)          3216      
__________

###  <font color=2214AF >  $\;\;$訓練神經網路</font>

In [10]:
model.fit(x_train, y_train, batch_size=100, epochs=20);

Instructions for updating:
Use tf.cast instead.
Epoch 1/20
60000/60000 [==============================] - 6s 94us/step - loss: 0.0898 - acc: 0.1519
Epoch 2/20
60000/60000 [==============================] - 4s 69us/step - loss: 0.0891 - acc: 0.2488
Epoch 3/20
60000/60000 [==============================] - 4s 69us/step - loss: 0.0862 - acc: 0.3065
Epoch 4/20
60000/60000 [==============================] - 4s 69us/step - loss: 0.0729 - acc: 0.3755
Epoch 5/20
60000/60000 [==============================] - 4s 68us/step - loss: 0.0580 - acc: 0.5474
Epoch 6/20
60000/60000 [==============================] - 4s 69us/step - loss: 0.0449 - acc: 0.6692
Epoch 7/20
60000/60000 [==============================] - 4s 69us/step - loss: 0.0321 - acc: 0.7804
Epoch 8/20
60000/60000 [==============================] - 4s 69us/step - loss: 0.0266 - acc: 0.8217
Epoch 9/20
60000/60000 [==============================] - 4s 69us/step - loss: 0.0233 - acc: 0.8454
Epoch 10/20
60000/60000 [===========================

###  <font color=2214AF >  $\;\;$結果測試</font>

In [11]:
score = model.evaluate(x_test, y_test)
print(f'測試資料的 loss: {score[0]:.5f}')
print(f'測試資料的正確率: {score[1]}')

10000/10000 [==============================] - 1s 73us/step
測試資料的 loss: 0.00615
測試資料的正確率: 0.9601


## <font color=FC8600> MODEL 1 : 將三層CNN中的filter數目(4,8,16)調整成(8,16,32) 其餘參數不變</font> 

In [0]:
model_1 = Sequential()

model_1.add(Conv2D(8, (5, 5), padding='same', input_shape=(28, 28, 1)))
model_1.add(Activation('relu'))

model_1.add(MaxPool2D(pool_size=(2,2)))

model_1.add(Conv2D(16, (5, 5), padding='same'))
model_1.add(Activation('relu'))

model_1.add(MaxPool2D(pool_size=(2,2)))

model_1.add(Conv2D(32, (5, 5), padding='same'))
model_1.add(Activation('relu'))

model_1.add(MaxPool2D(pool_size=(2,2)))

model_1.add(Flatten())
model_1.add(Dense(9))
model_1.add(Activation('relu'))

model_1.add(Dense(10))
model_1.add(Activation('softmax'))

In [13]:
model_1.compile(loss='mse', optimizer=SGD(lr=0.07), metrics=['accuracy'])
model_1.summary()
model_1.fit(x_train, y_train, batch_size=100, epochs=20);

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_4 (Conv2D)            (None, 28, 28, 8)         208       
_________________________________________________________________
activation_6 (Activation)    (None, 28, 28, 8)         0         
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 14, 14, 8)         0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 14, 14, 16)        3216      
_________________________________________________________________
activation_7 (Activation)    (None, 14, 14, 16)        0         
_________________________________________________________________
max_pooling2d_5 (MaxPooling2 (None, 7, 7, 16)          0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 7, 7, 32)          12832     
__________

In [14]:
score = model_1.evaluate(x_test, y_test)
print(f'測試資料的 loss: {score[0]:.5f}')
print(f'測試資料的正確率: {score[1]}')

10000/10000 [==============================] - 1s 83us/step
測試資料的 loss: 0.00455
測試資料的正確率: 0.9707


## <font color=FC8600> MODEL 2 : 將三層CNN中的filter數大小(5,5)調整成(3,3) 其餘參數不變</font> 

In [0]:
model_2 = Sequential()

model_2.add(Conv2D(4, (3, 3), padding='same', input_shape=(28, 28, 1)))
model_2.add(Activation('relu'))

model_2.add(MaxPool2D(pool_size=(2,2)))

model_2.add(Conv2D(8, (3, 3), padding='same'))
model_2.add(Activation('relu'))

model_2.add(MaxPool2D(pool_size=(2,2)))

model_2.add(Conv2D(16, (3, 3), padding='same'))
model_2.add(Activation('relu'))

model_2.add(MaxPool2D(pool_size=(2,2)))

model_2.add(Flatten())
model_2.add(Dense(9))
model_2.add(Activation('relu'))

model_2.add(Dense(10))
model_2.add(Activation('softmax'))

In [16]:
model_2.compile(loss='mse', optimizer=SGD(lr=0.07), metrics=['accuracy'])
model_2.summary()
model_2.fit(x_train, y_train, batch_size=100, epochs=20);

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_7 (Conv2D)            (None, 28, 28, 4)         40        
_________________________________________________________________
activation_11 (Activation)   (None, 28, 28, 4)         0         
_________________________________________________________________
max_pooling2d_7 (MaxPooling2 (None, 14, 14, 4)         0         
_________________________________________________________________
conv2d_8 (Conv2D)            (None, 14, 14, 8)         296       
_________________________________________________________________
activation_12 (Activation)   (None, 14, 14, 8)         0         
_________________________________________________________________
max_pooling2d_8 (MaxPooling2 (None, 7, 7, 8)           0         
_________________________________________________________________
conv2d_9 (Conv2D)            (None, 7, 7, 16)          1168      
__________

In [17]:
score = model_2.evaluate(x_test, y_test)
print(f'測試資料的 loss: {score[0]:.5f}')
print(f'測試資料的正確率: {score[1]}')

10000/10000 [==============================] - 1s 74us/step
測試資料的 loss: 0.00905
測試資料的正確率: 0.9407


## <font color=FC8600> MODEL 3 : 將三層CNN後的pooling size改為(3,3) 其餘參數不變</font> 

In [0]:
model_3 = Sequential()

model_3.add(Conv2D(4, (5, 5), padding='same', input_shape=(28, 28, 1)))
model_3.add(Activation('relu'))
model_3.add(MaxPool2D(pool_size=(3,3)))

model_3.add(Conv2D(8, (5, 5), padding='same'))
model_3.add(Activation('relu'))
model_3.add(MaxPool2D(pool_size=(3,3)))

model_3.add(Conv2D(16, (5, 5), padding='same'))
model_3.add(Activation('relu'))
model_3.add(MaxPool2D(pool_size=(3,3)))

model_3.add(Flatten())
model_3.add(Dense(9))
model_3.add(Activation('relu'))

model_3.add(Dense(10))
model_3.add(Activation('softmax'))

In [19]:
model_3.compile(loss='mse', optimizer=SGD(lr=0.07), metrics=['accuracy'])
model_3.summary()
model_3.fit(x_train, y_train, batch_size=100, epochs=20);

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_10 (Conv2D)           (None, 28, 28, 4)         104       
_________________________________________________________________
activation_16 (Activation)   (None, 28, 28, 4)         0         
_________________________________________________________________
max_pooling2d_10 (MaxPooling (None, 9, 9, 4)           0         
_________________________________________________________________
conv2d_11 (Conv2D)           (None, 9, 9, 8)           808       
_________________________________________________________________
activation_17 (Activation)   (None, 9, 9, 8)           0         
_________________________________________________________________
max_pooling2d_11 (MaxPooling (None, 3, 3, 8)           0         
_________________________________________________________________
conv2d_12 (Conv2D)           (None, 3, 3, 16)          3216      
__________

In [20]:
score = model_3.evaluate(x_test, y_test)
print(f'測試資料的 loss: {score[0]:.5f}')
print(f'測試資料的正確率: {score[1]}')

10000/10000 [==============================] - 1s 77us/step
測試資料的 loss: 0.01921
測試資料的正確率: 0.8652


## <font color=FC8600> MODEL 4 : 將三層CNN改為四層且filter數目為32 其餘參數不變</font> 

In [0]:
model_4 = Sequential()

model_4.add(Conv2D(4, (5, 5), padding='same', input_shape=(28, 28, 1)))
model_4.add(Activation('relu'))
model_4.add(MaxPool2D(pool_size=(2,2)))

model_4.add(Conv2D(8, (5, 5), padding='same'))
model_4.add(Activation('relu'))
model_4.add(MaxPool2D(pool_size=(2,2)))

model_4.add(Conv2D(16, (5, 5), padding='same'))
model_4.add(Activation('relu'))
model_4.add(MaxPool2D(pool_size=(2,2)))

model_4.add(Conv2D(32, (5, 5), padding='same'))
model_4.add(Activation('relu'))
model_4.add(MaxPool2D(pool_size=(2,2)))

model_4.add(Flatten())
model_4.add(Dense(9))
model_4.add(Activation('relu'))

model_4.add(Dense(10))
model_4.add(Activation('softmax'))

In [22]:
model_4.compile(loss='mse', optimizer=SGD(lr=0.07), metrics=['accuracy'])
model_4.summary()
model_4.fit(x_train, y_train, batch_size=100, epochs=20);

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_13 (Conv2D)           (None, 28, 28, 4)         104       
_________________________________________________________________
activation_21 (Activation)   (None, 28, 28, 4)         0         
_________________________________________________________________
max_pooling2d_13 (MaxPooling (None, 14, 14, 4)         0         
_________________________________________________________________
conv2d_14 (Conv2D)           (None, 14, 14, 8)         808       
_________________________________________________________________
activation_22 (Activation)   (None, 14, 14, 8)         0         
_________________________________________________________________
max_pooling2d_14 (MaxPooling (None, 7, 7, 8)           0         
_________________________________________________________________
conv2d_15 (Conv2D)           (None, 7, 7, 16)          3216      
__________

In [23]:
score = model_4.evaluate(x_test, y_test)
print(f'測試資料的 loss: {score[0]:.5f}')
print(f'測試資料的正確率: {score[1]}')

10000/10000 [==============================] - 1s 90us/step
測試資料的 loss: 0.01680
測試資料的正確率: 0.8768


## <font color=FC8600> MODEL 5 : 將NN的dense層改為20 其餘參數不變</font> 

In [0]:
model_5 = Sequential()

model_5.add(Conv2D(4, (5, 5), padding='same', input_shape=(28, 28, 1)))
model_5.add(Activation('relu'))
model_5.add(MaxPool2D(pool_size=(2,2)))

model_5.add(Conv2D(8, (5, 5), padding='same'))
model_5.add(Activation('relu'))
model_5.add(MaxPool2D(pool_size=(2,2)))

model_5.add(Conv2D(16, (5, 5), padding='same'))
model_5.add(Activation('relu'))
model_5.add(MaxPool2D(pool_size=(2,2)))

model_5.add(Flatten())
model_5.add(Dense(30))
model_5.add(Activation('relu'))

model_5.add(Dense(10))
model_5.add(Activation('softmax'))

In [25]:
model_5.compile(loss='mse', optimizer=SGD(lr=0.07), metrics=['accuracy'])
model_5.summary()
model_5.fit(x_train, y_train, batch_size=100, epochs=20);

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_17 (Conv2D)           (None, 28, 28, 4)         104       
_________________________________________________________________
activation_27 (Activation)   (None, 28, 28, 4)         0         
_________________________________________________________________
max_pooling2d_17 (MaxPooling (None, 14, 14, 4)         0         
_________________________________________________________________
conv2d_18 (Conv2D)           (None, 14, 14, 8)         808       
_________________________________________________________________
activation_28 (Activation)   (None, 14, 14, 8)         0         
_________________________________________________________________
max_pooling2d_18 (MaxPooling (None, 7, 7, 8)           0         
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 7, 7, 16)          3216      
__________

In [26]:
score = model_5.evaluate(x_test, y_test)
print(f'測試資料的 loss: {score[0]:.5f}')
print(f'測試資料的正確率: {score[1]}')

10000/10000 [==============================] - 1s 82us/step
測試資料的 loss: 0.00491
測試資料的正確率: 0.9683


## <font color=FC8600> MODEL 6 : 將loss function 改為'binary_crossentropy'其餘參數不變</font> 

In [0]:
model_6 = Sequential()

model_6.add(Conv2D(4, (5, 5), padding='same', input_shape=(28, 28, 1)))
model_6.add(Activation('relu'))
model_6.add(MaxPool2D(pool_size=(2,2)))

model_6.add(Conv2D(8, (5, 5), padding='same'))
model_6.add(Activation('relu'))
model_6.add(MaxPool2D(pool_size=(2,2)))

model_6.add(Conv2D(16, (5, 5), padding='same'))
model_6.add(Activation('relu'))
model_6.add(MaxPool2D(pool_size=(2,2)))

model_6.add(Flatten())
model_6.add(Dense(9))
model_6.add(Activation('relu'))

model_6.add(Dense(10))
model_6.add(Activation('softmax'))

In [28]:
model_6.compile(loss='binary_crossentropy', optimizer=SGD(lr=0.07), metrics=['accuracy'])
model_6.summary()
model_6.fit(x_train, y_train, batch_size=100, epochs=20);

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_20 (Conv2D)           (None, 28, 28, 4)         104       
_________________________________________________________________
activation_32 (Activation)   (None, 28, 28, 4)         0         
_________________________________________________________________
max_pooling2d_20 (MaxPooling (None, 14, 14, 4)         0         
_________________________________________________________________
conv2d_21 (Conv2D)           (None, 14, 14, 8)         808       
_________________________________________________________________
activation_33 (Activation)   (None, 14, 14, 8)         0         
_________________________________________________________________
max_pooling2d_21 (MaxPooling (None, 7, 7, 8)           0         
_________________________________________________________________
conv2d_22 (Conv2D)           (None, 7, 7, 16)          3216      
__________

In [29]:
score = model_6.evaluate(x_test, y_test)
print(f'測試資料的 loss: {score[0]:.5f}')
print(f'測試資料的正確率: {score[1]}')

10000/10000 [==============================] - 1s 88us/step
測試資料的 loss: 0.01373
測試資料的正確率: 0.9951199964523315


## 正確率比較
###課堂model ->96..01%
###model_1  ->97.07%
###model_2  ->94.07%
###model_3  ->86.52%
###model_4  ->87.68%
###model_5  ->96.83%
###model_6  ->99.51%
### 可以發現model_6調整loss function的表現最佳，測資正確率高達99.51%